### ISSUE TITLE
dagger: dind "failed to get digest" errors

### ISSUE THREAD

**Issue by alafanechere** [airbytehq/airbyte#27233]
On nightly build on which we disabled conccurency for java connectors (see https://github.com/airbytehq/airbyte/issues/27168#issuecomment-1584608324) 
we face errors from the global `dockerd` service:
```
2023-06-09T13:30:59.5034536Z #16 2312.7 time="2023-06-09T13:29:15.008451175Z" level=error msg="invalid image access: \"sha256:b5fadaa0074623dc614a16fb29df59277f57e96b0bdc5a23d6d5058f14e96172\", error: \"failed to get digest sha256:b5fadaa0074623dc614a16fb29df59277f57e96b0bdc5a23d6d5058f14e96172: open /var/lib/docker/image/overlay2/imagedb/content/sha256/b5fadaa0074623dc614a16fb29df59277f57e96b0bdc5a23d6d5058f14e96172: no such file or directory\""
```

Context:
* We initialize a [global `dockerd` service ](https://github.com/airbytehq/airbyte/blob/f8a9c3ed2b63acf2f5c663f08cfacc82c7202b07/tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py#L284) that is bound to containers that require docker interaction
* We mount a cache volume to `/var/lib/docker` to this dockerd service to get a persistent volume in case the dagger kills the service after the 10 seconds inactivity grace period.

**Comment by alafanechere** [comment]
Hypothesis:
**A. The cache volume has grown too big and some of its data was pruned.**
I don't think this hypothesis is valid because I tried mapping new volumes to `/var/lib/docker` with the same error occuring.

**B. We're hitting a docker bug that has been solved in more recent versions.**
I'm[ trying out docker 24 images](https://github.com/airbytehq/airbyte/actions/runs/5242639625/jobs/9466292855) for dind and cli and check if it has an effect.

**C. The dockerd service is not gracefully killed by Dagger and leads to corrupted volumes** 
I [enabled debug log level ](https://github.com/airbytehq/airbyte/actions/runs/5242639625/jobs/9466292855)for dockerd to get more insights on the dockerd service lifecyle.

**Comment by vito** [comment]
It looks like [`dockerd` gets restarted](https://alpha.dagger.cloud/runs/dbbf08fd-73f4-452b-8919-daab2a294e28#0-12243) before the error occurs (2023-06-12T13:28:24.8740475Z according to the logs from https://github.com/airbytehq/airbyte/issues/27233#issuecomment-1587460494). I'm guessing that's related.

To avoid the 10 second grace period and leave `dockerd` running the entire time, you could try directly running the service container with `container.sync` in the background. `container.sync` is a new API with similar behavior to `container.exit_code`, so if you don't have that yet, you can use `container.exit_code`. I'm not sure how to do this in Python, but in Go this would look like `go container.Sync(ctx)`.

**Comment by alafanechere** [comment]
Thank you @vito for the suggestion!
How can I run it in the background? If it's like `exit_code` I need to await it and it will be blocking the rest of the execution right?

Here's the function that currently createst the `dockerd` container.
```python
def with_global_dockerd_service(dagger_client: Client) -> Container:
    """Create a container with a docker daemon running.
    We expose its 2375 port to use it as a docker host for docker-in-docker use cases.
    Args:
        dagger_client (Client): The dagger client used to create the container.
    Returns:
        Container: The container running dockerd as a service
    """
    return (
        dagger_client.container()
        .from_(consts.DOCKER_DIND_IMAGE)
        .with_mounted_cache(
            "/var/lib/docker",
            dagger_client.cache_volume("docker-lib"),
            sharing=CacheSharingMode.SHARED,
        )
        .with_mounted_cache(
            "/tmp",
            dagger_client.cache_volume("shared-tmp"),
        )
        .with_exposed_port(2375)
        .with_exec(["dockerd", "--log-level=error", "--host=tcp://0.0.0.0:2375", "--tls=false"], insecure_root_capabilities=True)

    )
```

And I bound it to other containers as a service with:
````python

def with_bound_docker_host(
    context: ConnectorContext,
    container: Container,
) -> Container:
    """Bind a container to a docker host. It will use the dockerd service as a docker host.

    Args:
        context (ConnectorContext): The current connector context.
        container (Container): The container to bind to the docker host.
        shared_volume (Optional, optional): A tuple in the form of (mounted path, cache volume) that will be both mounted to the container and the dockerd container. Defaults to None.
        docker_service_name (Optional[str], optional): The name of the docker service, useful context isolation. Defaults to None.

    Returns:
        Container: The container bound to the docker host.
    """
    dockerd = context.dockerd_service
    docker_hostname = "global-docker-host"
    bound = (
        container.with_env_variable("DOCKER_HOST", f"tcp://{docker_hostname}:2375")
        .with_service_binding(docker_hostname, dockerd)
        .with_mounted_cache("/tmp", context.dagger_client.cache_volume("shared-tmp"))
    )
````

**Comment by alafanechere** [comment]
I confirmed that a [nightly build](https://github.com/airbytehq/airbyte/actions/runs/5252723316) on all our GA connectors succeeded without transient failure on #27267 
I also ran a[ nightly build for GA + Beta Java connectors](https://github.com/airbytehq/airbyte/actions/runs/5247814879) only with success.
I merged #27267 and triggered a [full nightly build on master.](https://github.com/airbytehq/airbyte/actions/runs/5254757865/jobs/9493768024). I'll close this issue if the error does not occur on this workflow.

**Comment by alafanechere** [comment]
This error did not occur in our latest nightly build following the merge of #27267  🎉
